### ISSUE TITLE
Problem trying to run XbimGeometryEngine

### ISSUE THREAD

**Issue by martin1cerny** [xBimTeam/XbimGeometry#2]
_From @IFChaos on June 11, 2015 18:13_

Hello, i'm new in Xbim, i need to load the geometry of an IFC file, but i obtain this error.

"Could not load file or assembly 'Xbim.Geometry.Engine32.dll' or one of its dependencies. The specified module could not be found."

The same problem emerges when i try to open a IFC file using XbimXplorer

Any ideas on what i'm doing wrong?

_Copied from original issue: xBimTeam/XbimWebUI#16_

**Comment by martin1cerny** [comment]
_From @LloydPickering on June 11, 2015 18:28_

The Xbim Geometry Nuget package includes 2 unmanaged C++ dlls for geometry generation (Engine 32 and Engine 64). These are loaded when creating an Xbim file from an IFC file, and need to be in a subdirectory in your Working Dir/Bin folder.

Restoring the package should ultimately run the following Powershell Script to insert the files into your Solution and copy them on build, but if it does not you need to do this manually:
https://github.com/xBimTeam/XbimGeometry/blob/master/install.ps1

**Comment by martin1cerny** [comment]
_From @IFChaos on June 11, 2015 19:4_

I copied Xbim.Geometry.Engine32 and Xbim.Geometry.Engine64 files in the BIN directory, but it still fails, i copied it in "Bin/Debug" also but still fails. I tried the same when running XbimXplorer, but anyway it fails. I tried it using folders X86 and X64 but it fails again.

I'm using a windows 7 x64 and Visual studio 2012. Maybe it's my computer that fails.

I'm lost at this point because i need to obtain the point coordinates and the triangles of columns, walls, etc.

**Comment by martin1cerny** [comment]
This whole issue thread should really be moved/asked in XbimGeometry repository as it is not related to XbimWebUI. Try to use NuGet distribution please. It should sort out all dependencies including geometry engine DLLs being placed to the right directory for you.

**Comment by martin1cerny** [comment]
_From @IFChaos on June 15, 2015 9:32_

I'm sorry, i posted it in the wrong place, Is there a method to move it to XbimGeometry or i have to close it and open a new one?

**Comment by martin1cerny** [comment]
I'll try to move it there.

**Comment by IFChaos** [comment]
Thank you for moving this issue and sorry for the inconvenience.

 At the moment i'm trying to obtain geometry from Xbim, i'm using the last updates and i place xbimgeometryEngine32 and 64 in the correct place. Anyway the error message keep showing saying that engine32 failed to load.
I think that it's not exactly xbimgeometryengine32 that fails to load, but one of it's dependencies.

As i need a quick solution i will try combine Xbim with a method i developed to obtain geometry from IFC
files.

Anyway if you know a method to obtain geometry without use xbim.geometry.engine it will be great to know it.

Thank you anyway.

**Comment by martin1cerny** [comment]
If you need a geometry only as a triangles and points you do need the geometry engine. Xbim.Geometry.Engine64.dll and Xbim.Geometry.Engine32.dll don't have any additional dependencies so it loads if it is found. If you use NuGet to add geometry engine to your solution it should just work. We have many different deployments running this way on desktops, webservers, Azure cloud and other environments.

**In case this is new for you:** *Right click on the solution folder in Solution Explorer -> Manage NuGet Packages for Solution -> search for "xbim" -> Install XbimGeometry. It will bring in all dependencies an will set up geometry engine dlls for you.*

**Comment by IFChaos** [comment]
Yes i follow the instructions to install nutget and i found the Xbim Geometry nutget and i installed it, but for some reason it fails to load.
Maybe the problem is that i'm using DIRECTX 9.0. for graphic representation, and i see that Xbim uses OpenGl. DirectX 9.0 does not admit Framework 4.5, but it is possible to use Framework 4.5 in combination with DirectX if you add "useLegacyV2RuntimeActivationPolicy="true"" in the App.Config.

This trick allowed me to install NutGet to my DirectX CAD application but maybe it is what is causing the whole problem.

I know it is confusing but it's probably my fault, not Xbim problem. Essentially i want to use Xbim to obtain Nodes and triangles and then represent it using my 3D DirectX Methods.

 i tried to place manually Xbim.Geometry.Engine32 and 64 in multiple possible locations in BIN/DEBUG, using folders x86 and x64

**Comment by martin1cerny** [comment]
We just currently started to look at OpenGL implementation but that is not
a part of xBIM yet. We only use WebGL for web presentation and native WPF
controls for Windows presentation. However, geometry engine is independent
of any of these.
Dne 15. 6. 2015 12:12 "IFChaos" <notifications@github.com> napsal(a):

**Comment by IFChaos** [comment]
Then i'm 100% lost i have no idea why it fails.
It's important to notice that when i download XbimWindowsUI projects, the XbimXplorer runs normally, but when i try to open an IFC model the same error message shows up. If fails to load Xbim.geometry.Engine32. Maybe something is wrong in my computer, it's a windows 7 x64. I'm using Visual Studio 2012.

**Comment by LloydPickering** [comment]
Ah... VS2012 *might* be the problem. I think min requirement is 2013+.

It is possibly failing because it is looking for C++ runtimes from VS2013. You should be able to find them here: https://www.microsoft.com/en-us/download/details.aspx?id=40784

**Comment by SteveLockley** [comment]
Xbim can be compiled from source to .Net 3.5 and 4.0 but the Nuget distributions are .Net 4.5 and the corresponding c runtimes

**Comment by andyward** [comment]
Sounds like a likely candidate Lloyd. We're building with VC12 now, which won't be on a vanilla VS2012 install.

I'm going to take a look at the code in the XbimGeometryEngine custom assembly resolver anyway as it fails in other environments (Azure - possibly the same problem). 

At the very least I'll try to add some better diagnostics. Do you think it may be worth putting a sanity check in for the VC runtime pre-requisites before we try probing for the Geometry DLL? E.g. as per http://stackoverflow.com/a/14564243 ... but checking for **VC12**

**Comment by IFChaos** [comment]
I think i find a possible origin of this problem. Investigating the error i try to find what is the path where C# tries to load Engine32 and i find it.

![errormsg](https://cloud.githubusercontent.com/assets/12625508/8202314/25536516-14d9-11e5-9a05-7541ed3266d4.jpg)

The path C:\users\steve\... is not of my computer... maybe it is problem. It is somehow using an old path and fails to find Geometry.Engine.

**Comment by andyward** [comment]
I think that's just the path from the PDB, which is deployed in the Xbim.Geometry nuget package.  The assembly resolver won't be using that - just the debugger.

I suspect the issue you're seeing is caused by one of two things:

1. the relevant Xbim.Geometry.Engine[32|64].dll is not in the bin folder (it should be in x86 or x64 subfolder)
or
2. a dependency of Xbim.Geometry.Engine cannot be located. E.g. the C++ Runtime (VC12)

You could always test the 2nd case by building a test case and referencing the Engine DLL directly and constructing an object. 

If you're able to, I have a test Xbim.Geometry nuget package that has diagnostics in it you can try. Will package it up and come back to you.

**Comment by IFChaos** [comment]
I created a Form project to test the Dll, it works, even when the method does not exists inside the dll, it does not give any error message.

This is the code

    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
               
        [DllImport("C:\\x86\\Xbim.Geometry.Engine32.dll", SetLastError = true, CharSet = CharSet.Auto)]
        private static extern Int32 GetShortPathName(String path);


        private void Form1_Load(object sender, EventArgs e)
        {
            Int32 Res = GetShortPathName("C:\\x86\\Xbim.Geometry.Engine32.dll");
        }
    }

So the problem should be somewhere else.

Is there any method to obtain the path that Xbim uses to load Engine32?

**Comment by andyward** [comment]
You don't need to DllImport the assembly. It's a .NET dll (but I believe it contains some unmanaged DLLs as embedded content). Just add a vanilla reference (and use the XBim.Essentials nuget) and new up a XbimGeometryCreator:

`var x = new Xbim.Geometry.XbimGeometryCreator()`

If you need some diagnostics, make sure you've set up log4net in your config file (see some of the xbim examples for .config)

**Comment by IFChaos** [comment]
Maybe it's a newbie question but, what is vanilla and how i load it?

**Comment by IFChaos** [comment]
Ok, i open a new Windows Form project in VS2012, i installed ALL the nugets related to Xbim, i added a reference to Geometry.Engine32, and the code you pasted to me worked.

No error message.

**Comment by andyward** [comment]
Just by browsing to the DLL in 'Add References'. The DLL is not explicitly referenced when you add the Geometry nuget, so we're just trying to establish if you can reference it directly. 
![image](https://cloud.githubusercontent.com/assets/923699/8205695/cd059d18-14ea-11e5-95a5-ae572881360d.png)

If no luck can you try using this custom Xbim.Geometry nuget:

https://www.dropbox.com/sh/srhs5jfc1nwavqb/AADOvjwclPIGnxq5h27Rkbfna?dl=0

You need to set up a local nuget folder and drop the nupkg in.
![image](https://cloud.githubusercontent.com/assets/923699/8205826/022606c6-14ec-11e5-97cc-70458fed01d0.png)

Then make sure you enable pre-release nuget packages in the nuget Package Manager:
![image](https://cloud.githubusercontent.com/assets/923699/8205807/e8ac9caa-14eb-11e5-8adc-b24331b3f2e7.png)

This package is as per 3.0.23 but just has a sanity check for the C runtime and some extra diagnostics. To see the logging splice the log4net section from app.config into your current app.config, and it'll spit out a log file called Default.log in the bin folder

**Comment by IFChaos** [comment]
Ok i mistake, it's very strange and still fails.

Somehow it loads Engine32 as reference without problems, but when i run it, when i press the button that executes ->

"var x = new Xbim.Geometry.XbimGeometryCreator();"

It fails. It shows a message saying 

Could not load file or assembly 'Xbim.Geometry.Engine32.dll' or one of its dependencies. The specified module could not be found.

Now i will try to install the last XbimEssentials 3.0.24 and i will post the results.

**Comment by andyward** [comment]
That sounds like you don't have the C++ runtime. Can you try my test nuget packages above? The Geometry Package has extra diagnostics that may help diagnose things. I'm hoping it flags up the lack of VC12.

**Comment by IFChaos** [comment]
Thank you in advance for your help and patience.

Ok i downloaded the nuget package from dropbox, i decompress it. Then i initiated a new VS2012 windows forms application, in this project i created a button with the function

"var x = new Xbim.Geometry.XbimGeometryCreator();"

I installed the local nugets you send to me using dropbox. Then i added Xbim.Geometry.Engine32.dll as reference.

Then i went to App.config, i copied the log4net part from the App.config you sent to me. But it fired a error saying that "name" attribute was not recognized, i searched it in google and solved it opening XML menu and creating a Schema.

Then i run it. I press the button and here is what i obtain:

![errormsg](https://cloud.githubusercontent.com/assets/12625508/8206932/e55e9bb8-14fd-11e5-97ba-ad378b2e9342.jpg)

I look for a log file in the Bin folder, but i find nothing. I think i'm doing something really wrong, but i have no idea on what. The only solution i find for this problem is a team viewer.

Anyway thank you in advance!

**Comment by andyward** [comment]
In terms of app.config, you should just be able to replace the whole lot. That should give you logging.

Can you show the contents of your bin folder (and bin/x86)? Also the salient code in Form1.cs? 

It looks like you're invoking something in Xbim.Geometry.Engine.**Interop** ?

**Comment by LloydPickering** [comment]
Andy, I think that sanity check is going to be a smart thing to add if only so we can get some sane error message to pop up so we can tell people to install the VC12 runtime.

**Comment by andyward** [comment]
Lloyd, I've put something in locally to do that. I was hoping this would test it out before I commit to github. 

@IFChaos Can you download this application and run it: https://www.dropbox.com/s/bqcmedgixec9ahn/Xbim.TestApp.zip?dl=0

It should just be a matter of compiling and running - I've included the packages and logging. If successful it will list the geometries of a simple model. If it fails (as I expect it to) please confirm the error.

Thanks

**Comment by IFChaos** [comment]
Hi Andy, yes this test fails giving this message "Could not load file or assembly 'Xbim.Geometry.Engine32' or one of its dependencies. An operation is not legal in the current state. (Exception from HRESULT: 0x80131509)"

Then i press F5 again and it produced this error:
"The Visual C++ runtime 'VC12' (VS2013) could not be located. This is required for XBim Geometry generation. Please install this from https://www.microsoft.com/en-us/download/details.aspx?id=40784"

So you were right !!! maybe this is the problem

I will try to install it i will tell you the result.

**Comment by IFChaos** [comment]
Ok i download the c++ runtime from the link and all worked perfectly.

Thank you

**Comment by LloydPickering** [comment]
Well done. I'll close the issue now.

**Comment by MarcoGiorgi** [comment]
I'm experiencing the same error but I'm using VS2015 and I have the VC14 (version: 14.0.24215.1) installed:

_Could not load file or assembly 'Xbim.Geometry.Engine64' or one of its dependencies. The system cannot find the file specified._

I have installed it with  NuGet, and everything is went fine:
_Attempting to gather dependency information for package 'Xbim.Geometry.4.0.11' with respect to project 'xBimtest', targeting '.NETFramework,Version=v4.6.1'
Gathering dependency information took 44.57 sec
Attempting to resolve dependencies for package 'Xbim.Geometry.4.0.11' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
Resolving actions to install package 'Xbim.Geometry.4.0.11'
Resolved actions to install package 'Xbim.Geometry.4.0.11'
Retrieving package 'Xbim.Geometry 4.0.11' from 'nuget.org'.
Adding package 'Xbim.Geometry.4.0.11' to folder 'E:\Progetti\temp\xBimTest\xBimtest\packages'
Added package 'Xbim.Geometry.4.0.11' to folder 'E:\Progetti\temp\xBimTest\xBimtest\packages'
Added package 'Xbim.Geometry.4.0.11' to 'packages.config'
Successfully installed 'Xbim.Geometry 4.0.11' to xBimtest
Executing nuget actions took 881.42 ms_

Using the code from the example on the website

const string fileName = "SampleHouse.ifc";
using (var model = IfcStore.Open(fileName))
      {
                var context = new Xbim3DModelContext(model);
                context.CreateContext();  // << HERE I GET THE ERROR!

                var wexBimFilename = Path.ChangeExtension(fileName, "wexBIM");
                using (var wexBiMfile = File.Create(wexBimFilename))
                {
                    using (var wexBimBinaryWriter = new BinaryWriter(wexBiMfile))
                    {
                        model.SaveAsWexBim(wexBimBinaryWriter);
                        wexBimBinaryWriter.Close();
                    }
                    wexBiMfile.Close();
        }
}

**Comment by martin1cerny** [comment]
Please, make sure Xbim.Geometry.Engine64.dll is present in your built/deployment. It is loaded in runtime based on the architecture and some deployment systems don't recognize it as a dependency.

**Comment by MarcoGiorgi** [comment]
Thanks, but **where can I find the Xbim.Geometry.Engine64.dll or Xbim.Geometry.Engine32.dll?**
In NuGet is not available.
Installing from NuGet the Xbim.Geometry (v4) it doesn't install it...

Additionally I've made a console application, on the same machine, using the same identical code, using the same identical references get from Nuget, and it works!

I'm wondering why it **doesn't work the WebSite project running on IIS** (both 32 or 64bit)

Thanks very much!

**Comment by MarcoGiorgi** [comment]
@martin1cerny Solved this way:
I've run the console app with **Xbim.Essential** and **Xbim.Geometry** libraries installed from NuGet.
At first compile it has "generated/written" the **Xbim.Geometry.EngineXX.dll** in the _bin/debug/x32_ and _bin/debug/x64_
I've copied the two folders, _x32_ and _x64_ in the bin folder of my website project.

Now it runs perfectly... and it seems wonderful!

**Comment by A-Shawqi** [comment]
Hello, I still have the same problem, I have a C# class library project that generates _ifc_ and _wexbim_ files using  XBIM Nugets (v.5). When I tried to call the the generated DLL from a console application, it thew exception "Failed to load Xbim.Geometry.Engine32.dll" but when I copied Xbim.Geometry.Engine32.dll and Xbim.Geometry.Engine64.dll to the debug folder of the console application (beside the exe not in the x86 and x64 folders) it works perfectly. 

The problem occurred when I tried to call my DLL from ASP.NET web project, the same exception was thrown:
![exception](https://user-images.githubusercontent.com/17937102/50542249-9a7d3000-0bc1-11e9-8662-463e128512b9.png)
 
I copied the Xbim.Geometry.EngineXX.dll in the bin/debug/x86 and bin/debug/x64 but it still fails and when i tried to copy them to the debug folder directly (same as the console application), I got the following error just after starting my project:
![error](https://user-images.githubusercontent.com/17937102/50542290-ab7a7100-0bc2-11e9-80b2-06f4d40bd84e.PNG)

Thank you in advance.
